core/models.py

```diff
@@ -0,0 +1,171 @@
+import hashlib
+import re
+from typing import ClassVar
+
+from django.contrib.auth import get_user_model
+from django.contrib.contenttypes.fields import GenericForeignKey, GenericRelation
+from django.contrib.contenttypes.models import ContentType
+from django.db import models
+
+User = get_user_model()
+
+
+class GenericModel(models.Model):
+    user = models.ForeignKey(User, on_delete=models.SET_NULL, null=True)
+    is_active = models.BooleanField(default=True)
+    is_locked = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    class Meta:
+        abstract = True
+
+
+class Community(GenericModel):
+    name = models.CharField(max_length=255)
+
+    def __str__(self: "Community") -> str:
+        return str(self.name)
+
+
+class Tag(models.Model):
+    name = models.SlugField()
+    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
+    object_id = models.PositiveIntegerField()
+    content_object = GenericForeignKey("content_type", "object_id")
+    created_at = models.DateTimeField(auto_now_add=True)
+
+    class Meta:
+        indexes: ClassVar[list[models.Index]] = [
+            models.Index(fields=["content_type", "object_id"]),
+        ]
+
+    def __str__(self: "Tag") -> str:
+        return str(self.name)
+
+
+class Image(models.Model):
+    image = models.ImageField()
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    def __str__(self: "Image") -> str:
+        return f"Image of {self.posts}"
+
+    def image_url(self: "Image") -> str:
+        try:
+            return self.image.url
+        except ValueError:
+            return ""
+
+
+class Post(GenericModel):
+    community = models.ForeignKey(
+        Community,
+        on_delete=models.CASCADE,
+        related_name="posts",
+    )
+    title = models.CharField(max_length=255)
+    content = models.TextField()
+    tags = GenericRelation(Tag, related_query_name="posts")
+    image = models.ManyToManyField(Image, blank=True, related_name="posts")

```

Comment:
Hmm I think. I would not make it many too many. 
Post can have multipole images, but image can have only one post so it should be foreign key.
On the image itself, we may need to have "ordering" field, int, 99999999. This way image can have ordering if needed and newly added will be always last.